# 🎓 Student Performance Prediction

## 📌 Problem Statement
Can we predict a student’s final grade (G3) using demographic, academic, and lifestyle features?
## 🎯 Objective
- Build and compare multiple ML models
- Identify key factors affecting student performance
- Evaluate model performance using R² score

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score 

## 📊 Dataset
- Source: UCI Machine Learning Repository
- Records: 395 students
- Features: 30+ variables (academic, social, personal)

In [4]:
data = pd.read_csv("student-mat.csv",sep=";")
data.head() 

,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,6,5,6,6
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,4,5,5,6
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,10,7,8,10
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,2,15,14,15
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,4,6,10,10


## Data Exploration
- Total rows and columns
- Data types
- Initial observations

In [5]:
data.shape
data.info()
data.describe()

<class 'pandas.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      395 non-null    str  
 1   sex         395 non-null    str  
 2   age         395 non-null    int64
 3   address     395 non-null    str  
 4   famsize     395 non-null    str  
 5   Pstatus     395 non-null    str  
 6   Medu        395 non-null    int64
 7   Fedu        395 non-null    int64
 8   Mjob        395 non-null    str  
 9   Fjob        395 non-null    str  
 10  reason      395 non-null    str  
 11  guardian    395 non-null    str  
 12  traveltime  395 non-null    int64
 13  studytime   395 non-null    int64
 14  failures    395 non-null    int64
 15  schoolsup   395 non-null    str  
 16  famsup      395 non-null    str  
 17  paid        395 non-null    str  
 18  activities  395 non-null    str  
 19  nursery     395 non-null    str  
 20  higher      395 non-null    str  
 21  inte

,age,Medu,Fedu,traveltime,studytime,failures,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
count,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000,395.000000
mean,16.696203,2.749367,2.521519,1.448101,2.035443,0.334177,3.944304,3.235443,3.108861,1.481013,2.291139,3.554430,5.708861,10.908861,10.713924,10.415190
std,1.276043,1.094735,1.088201,0.697505,0.839240,0.743651,0.896659,0.998862,1.113278,0.890741,1.287897,1.390303,8.003096,3.319195,3.761505,4.581443
min,15.000000,0.000000,0.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,3.000000,0.000000,0.000000
25%,16.000000,2.000000,2.000000,1.000000,1.000000,0.000000,4.000000,3.000000,2.000000,1.000000,1.000000,3.000000,0.000000,8.000000,9.000000,8.000000
50%,17.000000,3.000000,2.000000,1.000000,2.000000,0.000000,4.000000,3.000000,3.000000,1.000000,2.000000,4.000000,4.000000,11.000000,11.000000,11.000000
75%,18.000000,4.000000,3.000000,2.000000,2.000000,0.000000,5.000000,4.000000,4.000000,2.000000,3.000000,5.000000,8.000000,13.000000,13.000000,14.000000
max,22.000000,4.000000,4.000000,4.000000,4.000000,3.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,75.000000,19.000000,19.000000,20.000000


### Observations
- Both numerical and categorical features exist
- no missing values found


## ❓ Key Questions

- Which factors most influence student performance?
- Can lifestyle features predict academic success?
- Which model performs best on this dataset?

## Data Cleaning

### ⚠️ Data Leakage
G1(1st exam's grade) and G2(2nd exam's grade) were removed because they are previous exam scores and strongly correlated with G3(final grade).
Including them would result in unrealistically high model performance.

In [6]:
data = data.drop(['G1', 'G2'], axis=1)

### 🔄 Feature Preparation & Encoding

- Separated input (X) and target (y)
- Categorical variables were converted into numerical format using one-hot encoding. 
    This allows machine learning models to process non-numeric data.


In [12]:
X = data.drop('G3', axis=1)
y = data['G3']

X = pd.get_dummies(X)
### to convert boolean values to 0, 1
X = X.astype(int) 

After preprocessing, multiple regression models were trained and compared to evaluate their ability to predict student performance.

### Train-Test Split

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

## Model Training

### Linear Regression

In [16]:
lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)
r2_lr = r2_score(y_test, y_pred_lr)
r2_lr

0.20287177533488332

### 📈 Interpretation
Linear Regression performed poorly and achieved R² score of approximately 0.20.

This indicates weak linear relationship between features and target variable (G3). The model was able to capture only basic trends in the data and struggled with more complex relationships

### Decision Tree

In [17]:
dt = DecisionTreeRegressor(max_depth=3)
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)
r2_dt = r2_score(y_test, y_pred_dt)
r2_dt

0.2640411196788106

### 🌳 Interpretation

Decision Tree achieved the best performance with an R² score of approximately 0.39.

This suggests that the dataset contains non-linear relationships that tree-based models can capture more effectively than linear models. The model was able to learn useful decision rules from the student-related features.

### Random Forest

In [18]:
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
r2_rf = r2_score(y_test, y_pred_rf)
r2_rf

0.39017260113200525

### 🌲 Interpretation

Random Forest achieved an R² score of approximately 0.26.

Although Random Forest is generally more robust than a single Decision Tree, its performance in this case was lower. This may indicate that the dataset is relatively small and that averaging multiple trees reduced some useful patterns learned by the Decision Tree model.

## 📊 Overall Model Comparison

Among the evaluated models, Decision Tree achieved the highest R² score (~0.39), indicating that non-linear relationships exist within the dataset.

Linear Regression showed limited performance due to weak linear correlations between features and target values.

Random Forest improved upon Linear Regression but did not outperform the Decision Tree model in this case, possibly due to the relatively small dataset size and feature complexity.

In [20]:
results = pd.DataFrame({
    'Model': ['Linear Regression', 'Decision Tree', 'Random Forest'],
    'R2 Score': [r2_lr, r2_dt, r2_rf]
})

results

,Model,R2 Score
0,Linear Regression,0.202872
1,Decision Tree,0.264041
2,Random Forest,0.390173


## 🔍 Feature Importance Insights

Top features:
- absences → attendance impacts performance
- failures → strong negative academic signal
- goout → social activity affects study time
Surprisingly:
- studytime was not a top predictor, suggesting complexity in learning behavior

In [ ]:
feature_importance = pd.Series(rf.feature_importances_, index=X.columns)
feature_importance.sort_values(ascending=False).head(10).plot(kind='bar')
plt.title("Top Features")
plt.show()


## ⚠️ Limitations

- Dataset is small (395 samples)
- Many influencing factors (motivation, teaching quality) are not included
- Model performance is moderate (R² ≈ 0.33)

## 🚀 Future Work

- Try advanced models (XGBoost)
- Perform feature engineering
- Collect more meaningful features
- Use cross-validation for better evaluation

## ✅ Conclusion

- Random Forest achieved the best performance (R² ≈ 0.33)
- Academic history (failures) and behavior (absences, social activity) are key predictors
- Student performance is complex and not easily predicted using simple features
- This project helped strengthen understanding of data preprocessing, model evaluation, and interpretation of machine learning results using real-world educational data


## 📌 Key Takeaways

- Random Forest performed best (R² ≈ 0.33)
- Feature quality had greater impact than model choice
- Student performance is difficult to predict using limited features
- Proper preprocessing and leakage handling are essential in ML projects